src/CodeGeneration.jl

```diff
@@ -80,13 +80,14 @@ return_expression(::Array) = :(return result)
 
 """Should only be called if `all_constants(func_array) == true`. Unpredictable results otherwise."""
 function to_number(func_array::AbstractArray{T}) where {T<:Node}
+    _value(a::Node) = is_constant(a) ? value(a) : NaN
```

Comment:
@brianguenter I am not sure if you are happy with this fix since this implicitly promotes the array type to Float64 if there are non-constant expressions.

Reply:
@lassepe Thanks for taking the time to do this. I'm at a wedding and won't be back till Sunday evening so won't have much time to work on this till then. I'll look at it closely on Monday.

Reply:
How about this instead?

```
function to_number(func_array::AbstractArray{T}) where {T<:Node}
    #find type
    arr_type = typeof(value(func_array[begin]))
    for elt in func_array
        if is_constant(elt)
            arr_type = promote_type(arr_type, typeof(value(elt)))
        end
    end
    tmp = similar(func_array, arr_type)
    for (i, elt) in pairs(func_array)
        if is_constant(func_array[i])
            tmp[i] = value(func_array[i])
        end
    end
    return tmp
end
```
This makes the type of the array holding the constant values only dependent on the constants themselves without introducing NaN. All your new tests pass with this code.

Reply:
There are a few issues as outline below. I have added a few tests to highlight these challenges:

(1) If the first element is non-constant this will infer `arr_type` to be `Any` (or other inconsistent results) because `value` gives meaningless results for that first element. We could initialize `arr_type` to be the type of the first non-constant (or just initialize it to `Union{}` so that it promotes to the correct type once we hit the first constant). 

(2) **However**, I believe there is a fundamental problem with skipping non-constants: Imagine a function like this
```julia
x = make_variables(:x, 1)
f = Node.([2.1*x[1], 1, 2])
```
If we fix (1) by skipping non-constants entirely, we would infer the result type to be `Int` to which we cannot assign the Float64 result of (2.1*[x1]) as the code we generate would be
```julia
result = copy([140445214600784, 1, 2])
var"##260" = 2.1 * input_variables[1]
result[1] = var"##260"
return result
```

What we would need to do is to infer the return type of the non-constants as well so that the array has the correct promoted type that can hold *all* elements. When we skip the non-constant nodes, the tightest estimate we have is probably `Any` or `Number`. `Float64` is "correct" in most of the cases but even that would fail if the result is complex (though currently not supported) or the user tried to evaluate the generated function symbolically again.

Given these challenges, I currently can only think of the following solutions 

(a) Optionally allow the user to provide the return type. If they don't provide it, non-constants will be inferred to `Any`.  This is easy to implement but potentially a performance foot gun for the user. Then again, any user that really cares about speed would probably use the in-place version anyway where this is not an issue.

(b) Simply assume a wide worst-case return type that should be fine in "most cases"; e.g. `Float64`. This doesn't harm performance in many cases but makes the resultant functions less generic and also is not really sustainable if FD intends to support complex AD in the future (though maybe this latter case could be handled by checking if the *input* is `Complex`.

(c) Try harder to actually infer the return type of the non-constant nodes. This is probably the cleanest solution but writing an entire type inference engine for the computation graph may also be overkill (and hard to maintain) because we would have to effectively duplicate Julia internals.

Reply:
FWIW, the sparse-array code already assumes that the output type is `promote_type(Float64, eltype(input))`. This would be consistent with the `NaN` approach above (in that it lifts Ints to Float64).

Reply:
For now, I decided to promote the output type based on the input type for the non-inplace version. That seems to work very reliably; even with `Complex` or `Node` inputs to the generated functions.